### DISCUSSION TITLE
InvalidToken on an longstanding AWS Redash instance

### DISCUSSION THREAD

**Discussion by rankela** [getredash/redash#6831] [Q&A]
We have a running AWS EC2 instance created out of the Redash AMI that has been running for a year and a half with a few reboots. In the past couple of weeks we have been unable to perform queries or see and connect to data sources.

Upon examining the docker compose logs I have found plenty of InvalidToken exceptions thrown every time data sources are queried. I can paste the stack trace if needed. Quick search found that this is somehow related to the `REDASH_SECRET_KEY` environment variable, but it has been unchanged ever since the image was launched and as far I as understand it, it was autogenerated then.

Is there a way to mitigate this issue without a data loss, we have a large number of queries and dashboard that would be very painful to recreate.

Thanks!

**Comment by justinclift** [comment]
Well, first things first... back up the environment file (`/opt/redash/env`) and your Redash database.

With those two backed up safely, you can recreate your Redash installation if things go really wrong.

Unless there are further customisations that have been done to it, in which case you'll need to back those up too.

---

It *sounds* like restarting things isn't working for you.  Have you tried shutting down the containers completely (`docker-compose down` then `docker-compose up -d`)?  Sometimes that makes a difference with otherwise weird problems that make no sense.

If that doesn't work, then it sounds like there might really be some code level issue that needs investigating.  From your description of it running from the Redash AMI, it might be an older Redash version too.  Like, perhaps version 8 or similar?

If that's the case (after making a backup (not joking!)) you should try upgrading Redash using the process here:

https://redash.io/help/open-source/admin-guide/how-to-upgrade

The version numbers there are outdated (and should get fixed), so I'd try upgrading to version 10.1.0.  That uses the docker image labelled `10.1.0.b50633`.

Hopefully something of the above will fix the problem. :pray:

**Reply by rankela** [reply]
Hi, thanks for a prompt reply!

I have tried doing `docker-compose down` then `docker-compose up -d` and upgrading the version (it was indeed version 8) to 10.1.0 according to the instructions provided, but unfortunately it didn't help ☹️ 

When I try to list Data sources I now get Internal Server Error page (instead of an infinite spinner in version 8) and the error in logs is the same. I get the same message when trying to open any query or dashboard.

Any additional help would be greatly appreciated!

EDIT: Just wanted to add that I had to increase the instance RAM from 2GB to 4GB as it's suggested for V10, it was stuck with 100% CPU usage with lower RAM

**Reply by justinclift** [reply]
Ouch, that's not great.

Hmmm, as an experiment... what happens if you set up a brand new (eg fresh and empty) Redash installation (version 10.1.0) and then try connecting it to your data sources?

If that works, then (in theory) it might provide a reasonable base point to import the old configuration/queries/etc into (from the broken install).  Or at least provide a starting point to try and figure out WTF is going wrong. :wink:

If that also doesn't work, then I'm also out of ideas for the moment.  Someone else might have good suggestions though. :smile:

**Reply by rankela** [reply]
The experiment is a success, I made a new Redash server by starting from the AWS AMI and upgrading it to 10.1.0 and was able to connect it to my data source.

Regarding the configuration migration, how to deal with the encryption on original data?

Appreciate the help!

**Reply by justinclift** [reply]
Oops, sorry for not responding sooner.  Was rebuilding the test lab today (still going).

It's good news that the new Redash instance is playing ball.  In your situation, what I'd do is to run a temporary instance of your old Redash then export your data sources, queries, dashboards (etc) from that, then import them into the new (working) one.

And keep backups of everything just in case it turns out there's something in one of data sources or something that's *somehow* causing the problem.  aka just in case when you import the info... suddenly the new Redash starts breaking.

Does that make sense? :smile:

**Reply by justinclift** [reply]
> how to deal with the encryption on original data?

Actually, I'm not real clear on what you're asking there.  When you say "encryption on original data", are you meaning the queries in your original Redash instance?

Those should be encrypted using the `SECRET` values in your environment file (`/opt/redash/env`).  You *did* back that up right? :wink:

**Reply by rankela** [reply]
No worries, I really value your help!

How do I approach exporting the data? Do I export queries by dumping the related tables from the original PostgreSQL or is there maybe some tooling that could help me out?

The encryption question was asked assuming that's the case and that I should make a script that decrypts data with the original secret when doing the export and encrypt it on import with the new secret.

Also, everything is backed up and stored safely 😉

**Reply by justinclift** [reply]
You have a few options, depending on how good you are with Python and or writing code. :smile:

There's the official "redash-toolbelt", which you can use to extract queries and similar from a Redash instance:

https://github.com/getredash/redash-toolbelt

If you're ok with coding, but use Go... then you could use something like the old (3rd party) Go SDK:

https://github.com/RecoLabs/redash-go-sdk

It looks like they've archived that repo, but the Redash API hasn't really changed so it should be fine for extracting queries.

---

If none of that sounds like your kind of thing, then are you any good at using a terminal window and cli utilities like `curl`?

You can make curl calls directly to the Redash backend, and it'll return the info you're after.

Like this:

```
curl -s --request GET -H "Authorization: YOUR_API_KEY" -H 'Content-Type: application/json' https://YOUR_REDASH_SERVER/api/data_sources
```

That will return the raw info listing your data sources, in JSON format.

To make things pretty, it's common to add in the "jq" tool which automatically formats things to human readable output:

```
curl -s --request GET -H "Authorization: YOUR_API_KEY" -H 'Content-Type: application/json' https://YOUR_REDASH_SERVER/api/data_sources | jq
```

To retrieve info about a particular data source, append the data source id # on the end of the query.

For example, for data source 123 (they're all numbered in the previous output):

```
curl -s --request GET -H "Authorization: YOUR_API_KEY" -H 'Content-Type: application/json' https://YOUR_REDASH_SERVER/api/data_sources/123 | jq
```

The same concept applies to other things, like queries.  For example, for the list of all queries:

```
curl -s --request GET -H "Authorization: YOUR_API_KEY" -H 'Content-Type: application/json' https://YOUR_REDASH_SERVER/api/queries | jq
```

Then getting the details for a particular query (id 888 this time):

```
curl -s --request GET -H "Authorization: YOUR_API_KEY" -H 'Content-Type: application/json' https://YOUR_REDASH_SERVER/api/queries/888 | jq
```

---

That's for extracting info.  You can ***input*** data to the Redash server using curl `PUT` requests as well.

That's the approach the redash-toolbelt stuff above does, but uses Python instead of curl. :smile:

**Reply by rankela** [reply]
Thanks for a detailed answer!

I have managed to copy users, groups and queries over using `redash-migrate` and they seem to work. I have created data sources manually and added mapping to `meta.json` to enable migrating queries. Migrating dashboards results with an Internal server error (probably the underlying data source problem). Same thing happens when I try to curl GET any specific Dashboard, although I can GET the list of dashboards.

**Reply by rankela** [reply]
This inspired me to try something else -- On the NEW redash instance I ran `docker-compose run --rm server manage database reencrypt ${old_secret} ${new_secret}` where old_secret is the redash secret from the NEW redash instance env file and new_secret is from the OLD (non-functional) redash instance. Afterwards I connected to the redash psql databases and copied the contents of `encrypted_options` column in `data_sources` table. After this maneuver, the old server seems to be functioning again. I don't want to get my hopes up until I test it all, but I can see data sources and open queries and dashboards.

Thank you so much for your help, it was invaluable!

**Reply by justinclift** [reply]
Cool, sounds like you've got it pretty much figured out. :smile:
